c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +231,325 @@ struct Block {
+10 pages).  So we use 2MiB pages for our small pool and 20MiB pages for our
+large pool. Initially allocation using expandable_blocks will be slower than
+cudaMalloc, though still in the milliseconds range for mapping the entire
+memory.
+
+When mapping new memory to expand the segment, we look for the lowest address at
+which we can fit a new allocation by adding new pages. Normally this will be at
+the end of the block. But if have previously unmapped blocks earlier in the
+segment during an OOM, it will first try to fill in those gaps to keep the
+segment as a single block.
+
+Allocation of blocks in the segment uses the same best-fit heuristics of the
+rest of the allocator.
+
+Expandable blocks can be enabled/disabled throughout the run of a program. When
+disabled, the allocator will not put new allocations in an expandable block.
+
+Limitations
+
+* Slightly slower initial memory allocation speed.
+* IPC of cuda tensors (e.g. for multiprocess dataloaders) is not supported.
+However, it is possible to temporarily disable (expandable_segments:False) the
+bevhavior for allocator tensors that need to be used cross-process.
+* CUDA runtime APIs related to sharing memory across process
+(cudaDeviceEnablePeerAccess) do not work for memory allocated with cuMemMap.
+Instead these mapping have to be done manually. The allocator now has an
+`enablePeerAccess` method to do this.
+*/
+
+struct ExpandableSegment {
+  ExpandableSegment(
+      int device,
+      cudaStream_t stream,
+      size_t size,
+      const std::vector<int>& peers)
+      : device_(device),
+        stream_(stream),
+        max_handles_(0),
+        segment_size_(size),
+        peers_(peers) {
+    cudaDeviceProp prop;
+    C10_CUDA_CHECK(cudaGetDeviceProperties(&prop, device_));
+    max_handles_ = numSegments(prop.totalGlobalMem + prop.totalGlobalMem / 8);
+    C10_CUDA_DRIVER_CHECK(cuMemAddressReserve(
+        &ptr_, segment_size_ * max_handles_, 0ULL, 0, 0ULL));
+  }
+  // begin must be aligned to segment_size_.
+  // returns the actual range mapped, which may be
+  // greater than requested if size is not aligned to segment_size_.
+  // return size of 0 indicates OOM
+  SegmentRange map(SegmentRange range) {
+    auto begin = segmentLeft(range.ptr);
+    auto end = segmentRight(range.ptr + range.size);
+    TORCH_INTERNAL_ASSERT(ptr() + begin * segment_size_ == range.ptr);
+    if (begin == end) {
+      return rangeFromHandles(begin, end);
+    }
+    while (end > handles_.size()) {
+      handles_.push_back(c10::nullopt);
+    }
+    for (auto i : c10::irange(begin, end)) {
+      TORCH_INTERNAL_ASSERT(!handles_.at(i));
+      CUmemGenericAllocationHandle handle;
+      CUmemAllocationProp prop = {};
+      prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+      prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+      prop.location.id = device_;
+      auto status = cuMemCreate(&handle, segment_size_, &prop, 0);
+      if (status == CUDA_ERROR_OUT_OF_MEMORY) {
```

Comment:
What would happen from here if we're out of memory and return `rangeFromHandles` ?

Reply:
We end up returning a size 0 range `rangeFromHandles(begin, begin)` which basically says: you asked to map N bytes and we could only map 0. Before we return we also return any of the handles we were allocated in the call so that we reset the segment to as it was before the call happened.